lib/axon/loop.ex

```diff
@@ -866,6 +987,49 @@ defmodule Axon.Loop do
     %{loop | attached_state: state}
   end
 
+  @doc """
+  Serializes loop state to a binary for saving and loading
+  loop from previous states.
+
+  You can consider the serialized state to be a checkpoint of
+  all state at a given iteration and epoch.
+
+  By default, the step state is serialized using `Nx.serialize/2`;
+  however, this behavior can be changed if step state is an application
+  specific container. For example, if you introduce your own data
+  structure into step_state, `Nx.serialize/3` will not be sufficient
+  for serialization.
+
+  `opts` controls serialization options such as compression. It is
+  forwarded to `:erlang.term_to_binary/3`.
+  """
+  def serialize_state(%State{} = state, serialize_step_state_fn \\ &Nx.serialize/2, opts \\ []) do
+    serialized_step_state = serialize_step_state_fn.(state.step_state, opts)
+    serialized_metrics = Nx.serialize(state.metrics, opts)
+    state_map = Map.from_struct(state)
+    state_map = %{state_map | step_state: serialized_step_state, metrics: serialized_metrics}
+    :erlang.term_to_binary({@file_version, state_map}, opts)
+  end
+
+  @doc """
+  Deserializes loop state from a binary.
+
+  It is the opposite of `Axon.Loop.serialize_state/3`.
+
+  By default, the step state is deserialized using `Nx.deserialize.2`;
+  however, this behavior can be changed if step state is an application
+  specific container. For example, if you introduce your own data
+  structure into step_state and you customized the serialization logic,
+  `Nx.deserialize/2` will not be sufficient for deserialization.
+  """
+  def deserialize_state(serialized, deserialize_step_state_fn \\ &Nx.deserialize/2, opts \\ []) do
+    {1, state_map} = :erlang.binary_to_term(serialized, [:safe | opts])
+    step_state = deserialize_step_state_fn.(state_map.step_state, opts)
+    metrics = Nx.deserialize(state_map.metrics, opts)
+    state_map = %{state_map | step_state: step_state, metrics: metrics}
+    struct(Axon.Loop.State, state_map)
```

Comment:
```suggestion
    struct!(Axon.Loop.State, state_map)
```

So we raise on field mismatch?